### PR TITLE
Topic get species by name

### PR DESCRIPTION
**Pull request by psychocoderHPC**
solve #2393

Instead of passing the particle species type to a functor like `CreateDensity` it is possible to use the name of the species. This solve an issue if two species cyclical reference to each other.

- PMacc: new traits ￼…
  - add generic `GetCTName` trait to get a compile time name of a class
  - add compile time method to find a species type by name or by the type
  it self within a sequence of species types
- PIConGPU:
  - remove the usage of `MakeIdentifier` to pass species to functors
  - use `particles::compileTime::FindByNameOrType_t` to search for a species within `VectorAllSpecies`
  - remove include from `species.param` which is useless and break the compile

# Example

```C++
using InitPipeline = mpl::vector<
        CreateDensity<
            densityProfiles::Foil,
            startPosition::Quiet1ppc,
            PIC_Ions
        >,
        CreateDensity<
            densityProfiles::Foil,
            startPosition::Random100ppc,
            PIC_Electrons
        >
    >;
```

can be written now as
```C++
using Electrons = bmpl::string< 'e' >;
using Ions = bmpl::string< 'i' >;

using InitPipeline = mpl::vector<
        CreateDensity<
            densityProfiles::Foil,
            startPosition::Quiet1ppc,
            Ions
        >,
        CreateDensity<
            densityProfiles::Foil,
            startPosition::Random100ppc,
            Electrons
        >
    >;
```

CC-ing: @n01r

### CODE REVIEW

File: include/pmacc/particles/compileTime/FindByNameOrType.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+
+namespace pmacc
+{
+namespace particles
+{
+namespace compileTime
+{
+
+    /* find a type within a sequence by name or the type it self
```
Comment (ax3l) [inline include/pmacc/particles/compileTime/FindByNameOrType.hpp]: itself

File: include/pmacc/particles/compileTime/FindByNameOrType.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                >
+            >;
+        };
+
+        using FilteredSeq = typename bmpl::copy_if<
+            T_MPLSeq,
+            HasTypeOrName< bmpl::_1 >
+        >::type;
+
+        using type =typename bmpl::if_<
```
Comment (ax3l) [inline include/pmacc/particles/compileTime/FindByNameOrType.hpp]: style

File: include/pmacc/particles/compileTime/FindByNameOrType.hpp (hunk 1 above)
Comment (ax3l) [reply include/pmacc/particles/compileTime/FindByNameOrType.hpp]: you don't get the type by "a name" but specifically by the `pmacc::traits::GetCTName` trait. maybe we write that in the long description?

Comment (PrometheusPi) [inline include/pmacc/particles/compileTime/FindByNameOrType.hpp]: TYPO: itself

File: include/pmacc/traits/GetCTName.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    template< typename T_Type >
+    struct GetCTName
+    {
+        using type = boost::mpl::string< >;
+    };
+
+    template< typename T_Type >
+    using GetCTName_t = typename GetCTName< T_Type >::type;
+
+}//namespace traits
```
Comment (ax3l) [inline include/pmacc/traits/GetCTName.hpp]: style

File: include/picongpu/simulation_defines/param/species.param [hunk 4]
```diff
@@ -30,12 +30,8 @@
 #include "picongpu/algorithms/FieldToParticleInterpolationNative.hpp"
 #include "picongpu/algorithms/FieldToParticleInterpolation.hpp"
 #include "picongpu/algorithms/AssignedTrilinearInterpolation.hpp"
-
-#include "picongpu/fields/currentDeposition/Solver.def"
-
-#include "picongpu/particles/InitFunctors.hpp"
 #include "picongpu/particles/flylite/NonLTE.def"
-#include "picongpu/particles/manipulators/manipulators.def"
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/species.param]: @psychocoderHPC I think you forgot to update 4 examples with this change.

**Pull request merged**